module/zfs/dmu.c

```diff
@@ -1290,515 +1879,502 @@ dmu_write_uio_dnode(dnode_t *dn, uio_t *uio, uint64_t size, dmu_tx_t *tx)
-	 */
-	if (txg > spa_freeze_txg(os->os_spa))
-		return (dmu_sync_late_arrival(pio, os, done, zgd, &zp, &zb));
+	return (err);
+}
 
-	/*
-	 * Grabbing db_mtx now provides a barrier between dbuf_sync_leaf()
-	 * and us.  If we determine that this txg is not yet syncing,
-	 * but it begins to sync a moment later, that's OK because the
-	 * sync thread will block in dbuf_sync_leaf() until we drop db_mtx.
-	 */
-	mutex_enter(&db->db_mtx);
+/*
+ * Write 'size' bytes from the uio buffer.
+ * To the specified object.
+ * Starting at offset uio->uio_loffset.
+ */
+int
+dmu_write_uio(objset_t *os, uint64_t object, uio_t *uio, uint64_t size,
+    dmu_tx_t *tx)
+{
+	dnode_t *dn;
+	int err;
 
-	if (txg <= spa_last_synced_txg(os->os_spa)) {
-		/*
-		 * This txg has already synced.  There's nothing to do.
-		 */
-		mutex_exit(&db->db_mtx);
-		return (SET_ERROR(EEXIST));
-	}
+	if (size == 0)
+		return (0);
 
-	if (txg <= spa_syncing_txg(os->os_spa)) {
-		/*
-		 * This txg is currently syncing, so we can't mess with
-		 * the dirty record anymore; just write a new log block.
-		 */
-		mutex_exit(&db->db_mtx);
-		return (dmu_sync_late_arrival(pio, os, done, zgd, &zp, &zb));
-	}
+	err = dnode_hold(os, object, FTAG, &dn);
+	if (err)
+		return (err);
 
-	dr = dbuf_find_dirty_eq(db, txg);
+	err = dmu_write_uio_dnode(dn, uio, size, tx, B_FALSE);
 
-	if (dr == NULL) {
-		/*
-		 * There's no dr for this dbuf, so it must have been freed.
-		 * There's no need to log writes to freed blocks, so we're done.
-		 */
-		mutex_exit(&db->db_mtx);
-		return (SET_ERROR(ENOENT));
-	}
+	dnode_rele(dn, FTAG);
 
-	dr_next = list_next(&db->db_dirty_records, dr);
-	ASSERT(dr_next == NULL || dr_next->dr_txg < txg);
+	return (err);
+}
+#endif /* _KERNEL */
 
-	if (db->db_blkptr != NULL) {
-		/*
-		 * We need to fill in zgd_bp with the current blkptr so that
-		 * the nopwrite code can check if we're writing the same
-		 * data that's already on disk.  We can only nopwrite if we
-		 * are sure that after making the copy, db_blkptr will not
-		 * change until our i/o completes.  We ensure this by
-		 * holding the db_mtx, and only allowing nopwrite if the
-		 * block is not already dirty (see below).  This is verified
-		 * by dmu_sync_done(), which VERIFYs that the db_blkptr has
-		 * not changed.
-		 */
-		*zgd->zgd_bp = *db->db_blkptr;
-	}
+/*
+ * Allocate a loaned anonymous arc buffer.
+ */
+arc_buf_t *
+dmu_request_arcbuf(dmu_buf_t *handle, int size)
+{
+	dmu_buf_impl_t *db = (dmu_buf_impl_t *)handle;
+
+	return (arc_loan_buf(db->db_objset->os_spa, B_FALSE, size));
+}
+
+/*
+ * Free a loaned arc buffer.
+ */
+void
+dmu_return_arcbuf(arc_buf_t *buf)
+{
+	arc_return_buf(buf, FTAG);
+	arc_buf_destroy(buf, FTAG);
+}
+
+/*
+ * When possible directly assign passed loaned arc buffer to a dbuf.
+ * If this is not possible copy the contents of passed arc buf via
+ * dmu_write().
+ */
+int
+dmu_assign_arcbuf_by_dnode(dnode_t *dn, uint64_t offset, arc_buf_t *buf,
```

Comment:
Don't we need a Direct IO path here? This is a common IO path for Lustre.

Reply:
@behlendorf has added in a O_DIRECT path here for Lustre if the `direct` property is set to `always`.